### PR TITLE
JS FFI initial commit.

### PR DESCRIPTION
**Pull request by belmarca**
Working implementation of the JS FFI. The type conversion procedures are bound to evolve. An ergonomics layer can now be built on top.

### CODE REVIEW

File: include/lang/py/py.js [hunk 1]
```diff
@@ -813,3 +813,102 @@ function runtime_get_syntaxError_thrower(compilationError) {
...
+        return obj.seq.map(py2host);
+    }
+    // TODO: dicts
+    if (type(pyinterp.class_bool)) {
+        return obj.value && true;
+    }
+    if (type(pyinterp.class_NoneType)) {
+        return null
+    }
+}
```
Comment (feeley) [inline include/lang/py/py.js:873]: should have a "catch all" here... like raising an exception

File: include/lang/py/py.js [hunk 2]
```diff
@@ -813,3 +813,102 @@ function runtime_get_syntaxError_thrower(compilationError) {
...
+        return pyinterp.om_list(obj.map(host2py));
+    }
+    if ((obj === null) || (obj === undefined)) {
+        return pyinterp.om_None;
+    }
+    // This comes last to match only dicts
+    if (obj instanceof Object) {
+        return pyinterp.om_str("object conversion not implemented");
+    }
+}
```
Comment (feeley) [inline include/lang/py/py.js:902]: should have a "catch all" here... like raising an exception

File: include/lang/py/py.js [hunk 3]
```diff
@@ -813,3 +813,102 @@ function runtime_get_syntaxError_thrower(compilationError) {
...
+    }
+    if (type(pyinterp.class_NoneType)) {
+        return null
+    }
+}
+
+// Convert a JS object to a pyinterp object
+function host2py(obj) {
+    if (typeof obj === "boolean") {
+        return pyinterp.om_bool(true && obj);
```
[the remark this answers is not in the document]
Comment (feeley) [reply include/lang/py/py.js:878]: maybe `!!obj` is simpler

[the remark this answers is not in the document]
Comment (feeley) [reply include/lang/py/py.js:878]: actually we know `obj` is a boolean, so just `obj` will be fine

File: include/lang/py/py.js (hunk 1 above)
Comment (belmarca) [reply include/lang/py/py.js:874]: See `throw`.

File: include/lang/py/py.js (hunk 2 above)
Comment (belmarca) [reply include/lang/py/py.js:902]: I use `absent` as an indicator.

File: include/lang/py/py.js [hunk 4]
```diff
@@ -813,3 +813,105 @@ function runtime_get_syntaxError_thrower(compilationError) {
...
+
+// Convert a JS function to a pyinterp function
+function host_function2py(fn) {
+    var name = gen_foreign();
+    var signature = pyinterp.make_vararg_only_signature('args');
+    function code(rte, cont) {
+        var fn_args = py2host(pyinterp.rte_lookup_locals(rte, 'args'));
+        var result = host2py(fn.apply(null, fn_args));
+        if (result === pyinterp.absent) {
+            return pyinterp.sem_raise_with_message(pyinterp.make_out_of_ast_context(rte, cont), pyinterp.class_ValueError, "Could not convert JavaScript type to Python")
```
Comment (belmarca) [inline include/lang/py/py.js:912]: An exception must be raised from the returned procedure if its own return type cannot be converted.

File: pyinterp/pyinterp/pyinterp/__init__.py [hunk 5]
```diff
@@ -8603,6 +8603,14 @@ def inject_module_in_env(module, env):
         value = key_value[1]
         dict_set(env, key, value)
 
+# JS FFI
+def om_js_eval_code(rte, cont):
+    expr = rte_lookup_locals(rte, 'expr')
+    result = host2py(host_eval(py2host(expr)))
+    if result == absent:
+        return sem_raise_with_message(make_out_of_ast_context(rte, cont), class_ValueError, "Could not convert JavaScript type to Python")
```
Comment (belmarca) [inline pyinterp/pyinterp/pyinterp/__init__.py:8611]: `absent` indicates an unsupported conversion and must raise.

File: include/lang/py/py.js (hunk 1 above)
Comment (feeley) [reply include/lang/py/py.js:874]: where?

File: include/lang/py/py.js (hunk 2 above)
Comment (feeley) [reply include/lang/py/py.js:900]: shouldn't it be an explicit `return ...absent...;`?

**Review by feeley — APPROVED** [review]

**Pull request merged**